rust/worker/src/execution/operators/metadata_filtering.rs

```diff
@@ -0,0 +1,1348 @@
+                        tracing::error!("Error creating record segment reader {}", e);
+                        return Err(MetadataFilteringError::MetadataFilteringRecordSegmentReaderCreationError(
+                            RecordSegmentReaderCreationError::InvalidNumberOfFiles,
+                        ));
+                    }
+                };
+            }
+        };
+        // Step 0.5: Get the current max offset id for materialization.
+        // Offset Ids start from 1.
+        let mut curr_max_offset_id = Arc::new(AtomicU32::new(1));
+        match &record_segment_reader {
+            Some(reader) => {
+                curr_max_offset_id = reader.get_current_max_offset_id();
+                curr_max_offset_id.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+            }
+            None => (),
+        };
+        // Step 1: Materialize the logs.
+        let materializer = LogMaterializer::new(
+            record_segment_reader,
+            input.log_record.clone(),
+            curr_max_offset_id,
+        );
+        let mat_records = match materializer.materialize().await {
+            Ok(records) => records,
+            Err(e) => {
+                return Err(MetadataFilteringError::MetadataFilteringLogMaterializationError(e));
+            }
+        };
+        // Step 2: Apply where and where_document clauses on the materialized logs.
+        let mut ids_to_metadata: HashMap<u32, HashMap<&String, &MetadataValue>> = HashMap::new();
+        let mut ids_in_mat_log = HashSet::new();
+        for (records, _) in mat_records.iter() {
+            ids_in_mat_log.insert(records.offset_id);
+            // Skip deleted records.
+            if records.final_operation == Operation::Delete {
+                continue;
+            }
+            if !ids_to_metadata.contains_key(&records.offset_id) {
+                ids_to_metadata.insert(records.offset_id, HashMap::new());
+            }
+            let map_pointer = ids_to_metadata.get_mut(&records.offset_id).expect(
+                "Just inserted the key one line above so cannot happen that it does not exist now",
+            );
+            match &records.data_record {
+                Some(data_record) => match &data_record.metadata {
+                    Some(meta) => {
+                        for (meta_key, meta_val) in meta {
+                            map_pointer.insert(&meta_key, &meta_val);
+                        }
+                    }
+                    None => (),
+                },
+                None => (),
+            };
+            match &records.metadata_to_be_merged {
+                Some(meta) => {
+                    for (meta_key, meta_val) in meta {
+                        map_pointer.insert(meta_key, meta_val);
+                    }
+                }
+                None => (),
+            };
+        }
+        let clo = |metadata_key: &str,
```

Comment:
Would it be cleaner to make this a function elsehwere and just padd in the ids_to_metadata and any other captured values?